compiler/optimizer/abstractinterpreter/InliningMethodSummary.cpp

```diff
@@ -0,0 +1,166 @@
+      traceMsg(comp, "EMPTY\n\n");
+      return;
+      }
+
+   for (size_t i = 0; i < _optsByArg.size(); i ++) 
+      {
+      if (_optsByArg[i] != NULL)
+         {
+         for (size_t j = 0; j < _optsByArg[i]->size(); j ++)
+            {
+            TR::PotentialOptimizationPredicate* predicate = (*_optsByArg[i])[j];
+
+            traceMsg(comp, "%s @%d for Argument %d ", predicate->getName(), predicate->getBytecodeIndex(), i);
+            predicate->trace(comp);
+            traceMsg(comp, "\n");
+            }
+         }
+      }
+
+   }
+
+
+void TR::InliningMethodSummary::addPotentialOptimizationByArgument(TR::PotentialOptimizationPredicate* predicate, uint32_t argPos)
+   {
+   if (_optsByArg.size() <= argPos)
+      _optsByArg.resize(argPos + 1);
+      
+   if (!_optsByArg[argPos])
+       _optsByArg[argPos] = new (region()) PredicateContainer(region());
+      
+   _optsByArg[argPos]->push_back(predicate);
+   }
+
+const char* TR::PotentialOptimizationPredicate::getName()
+   {
+   switch (_kind)
+      {
+      case Kind::BranchFolding:
+         return "Branch Folding";
+      case Kind::CheckCastFolding:
+         return "CheckCast Folding";
+      case Kind::NullCheckFolding:
+         return "NullCheck Folding";
+      case Kind::InstanceOfFolding:
+         return "InstanceOf Folding";
+      default:
+         TR_ASSERT_FATAL(false, "Unexpected Kind");
+      }
+   }
+
+bool TR::PotentialOptimizationVPPredicate::holdPartialOrderRelation(TR::VPConstraint* valueConstraint, TR::VPConstraint* testConstraint)  
+   {
+   TR_ASSERT_FATAL(testConstraint->isClassObject() != TR_yes, "testConstraint is not a class Object");
+   TR_ASSERT_FATAL(testConstraint->getClass() != NULL, "testConstraint class is Null");
+   TR_ASSERT_FATAL(testConstraint->isNonNullObject(), "testConstraint is a Null");
+   TR_ASSERT_FATAL(testConstraint->getPreexistence() == NULL, "testConstraint already exists");
+   TR_ASSERT_FATAL(testConstraint->getArrayInfo() == NULL, "testConstraint Array Info is NUll");
+   TR_ASSERT_FATAL(testConstraint->getObjectLocation() == NULL, "testContraint object location is Null");
+   if (testConstraint->asIntConstraint()) // partial relation for int constraint
+      {
+      if (testConstraint->getLowInt() <= valueConstraint->getLowInt() && testConstraint->getHighInt() >= valueConstraint->getHighInt()) 
+         return true;
+      else 
+         return false;
+      }
+   else if (testConstraint->asClassPresence()) // partial relation for nullness
+      {
+      if (testConstraint->isNonNullObject() && valueConstraint->isNonNullObject())
+         return true;
+      else if (testConstraint->isNullObject() && valueConstraint->isNullObject())
+         return true;
+      else 
+         return false;
+      }
+   else if (testConstraint->asClassType()) // testing for checkcast
+      {
+       if (valueConstraint->getClassType()->asResolvedClass())
+         return !dynamic_cast<TR::VPConstraint*>(testConstraint);
```

Comment:
`testConstraint` will always be an instance of `VPConstraint`

This should still relate `testConstraint->getClass()` to `valueConstraint->getClass()`, just not by an equality comparison. Here we need `valueConstraint->getClass()` to be a subtype of `testConstraint->getClass()`. The subtype relationship can be determined using `TR_J9VM::isInstanceOf()`

Reply:
Since this is OMR code, I should have specified `TR_FrontEnd::isInstanceOf()` rather than `TR_J9VM`. I've edited my prior comment to correct that